solr/core/src/java/org/apache/solr/update/TransactionLog.java

```diff
@@ -987,4 +1022,28 @@ public String toString() {
           + position();
     }
   }
+
+  /** Opens {@link OutputStream} from {@link FileChannel}. */
+  protected interface OutputStreamOpener {
+
+    /**
+     * Opens an {@link OutputStream} to write in a {@link FileChannel}.
+     *
+     * @param position The initial write position of the {@link OutputStream} view of the {@link
+     *     FileChannel}.
+     */
+    OutputStream open(FileChannel channel, long position) throws IOException;
+  }
+
+  /** Opens {@link ChannelFastInputStream} from {@link FileChannel}. */
+  protected interface ChannelInputStreamOpener {
+
+    /**
+     * Opens a {@link ChannelFastInputStream} to read a {@link FileChannel}.
+     *
+     * @param position The initial read position of the {@link OutputStream} view of the {@link
+     *     FileChannel}.
+     */
+    ChannelFastInputStream open(FileChannel channel, long position) throws IOException;
```

Comment:
the previous impl (calling `new ChannelFastInputStream`) did not throw IOException. introducing this new signature changes the flow a bit and I was curious if this was necessary. as in, instead of adding `throws ...` just take care of the IOException in the new impl you want to introduce.
what I am concerned about is the below change (https://github.com/apache/solr/pull/2012/files#diff-754514ecac361b4d69e8e7c9948f59befe8986afe643562ccb7f192af923f666R1932-R1933) where now you can have an UncheckedIOException bubble up, instead of the previous behavior of simply logging.

Reply:
I added `throws IOException` to `getReader()` and `getSortedReader()` first for consistency with `getReverseReader()` which already throws IOE. I want the caller to know if there is an issue when creating a reader, and up to the caller to decide what to do.
Then, in `UpdateLog.doReplay()`, we have the choice of either catch and log and return, or throw an unchecked exception which is caught anyway in `UpdateLog.run()` to log it.

Reply:
ok I see what you are saying. it's all handled in `#run` method (this code is very confusing to read, some exceptions are handled locally some bubble up). I am ok with your changes, as you point out the exception is handled one level up so nothing changes in the flow.